src/libraries/System.Net.Http/src/System/Net/Http/HttpRequestMessage.cs

```diff
@@ -183,7 +185,28 @@ private void InitializeValues(HttpMethod method, Uri? requestUri)
 
         internal bool MarkAsSent()
         {
-            return Interlocked.Exchange(ref _sendStatus, MessageAlreadySent) == MessageNotYetSent;
+            return Interlocked.CompareExchange(ref _sendStatus, MessageAlreadySent, MessageNotYetSent) == MessageNotYetSent;
```

Comment:
Why did this need to be upgraded to a CompareExchange?

Reply:
If `_sendStatus` is set to `MessageShouldEmitTelemetry` and `MarkAsSent` is called, we would reset the status back to `MessageAlreadySent` without ever calling `RequestStop` for the request.

The Interlocked synchronization around the status ensures that `RequestStop` is called 1:1 to `MarkAsTrackedByTelemetry` calls.

Reply:
> If _sendStatus is set to MessageShouldEmitTelemetry and MarkAsSent is called

Won't that only happen if the request is being reused, in which case HttpClient is going to throw?

Reply:
> Won't that only happen if the request is being reused

Yes, MarkAsSent is the method used to guard against request re-use. While we will throw for the second request, having a regular Exchange here would mean we also suppress the `RequestStop` telemetry for the first request that is in progress.

Reply:
But the program is already invalid at that point. Losing the stop event for such invalid use seems fine.

Reply:
(if you measure and there's zero impact to the change, then ok, but I don't think it's worth any measurable cost)

Reply:
In a [microbenchmark](https://gist.github.com/MihaZupan/fd9e462c1840a595d14879929044fb33) on my machine there appears to be no regression (Compare being a fraction of a nanosecond faster)